### DISCUSSION TITLE
Upgrade from older GTK template

### DISCUSSION THREAD

**Discussion by jeromelaban** [unoplatform/uno#17571] [Q&A] [answered]
_from @hakanL in https://github.com/unoplatform/uno/issues/15738:_

@MartinZikmund Is there any documentation on how to upgrade from GTK to Desktop? I've tried to create a new application using the templates, but since my real application is like 2 years old, the differences are quite big (single project, etc) and it's hard to figure out a good starting point. Could I perhaps start with just changing GtkHost to SkiaHostBuilder?

**Comment by jeromelaban** [comment]
@hakanL Upgrading using the `SkiaHostBuilder` directly may not necessarily work properly on all platforms, because of the automatic redirection done in `net8.0-desktop`. That being said, migrating to the newer structure is mostly about removing code and/or configuration in csproj. See the [upgrade doc](https://platform.uno/docs/articles/migrating-to-single-project.html).

**Reply by HakanL** [reply]
Thanks, I'll review the upgrade doc, and see how it will play out with my solution.

**Reply by HakanL** [reply]
I was able to get my solution over to a single project file, but is this supported in Uno 5.1? I'm getting the following error when setting the Uno.Sdk to 5.1.104. (The reason I can't go to 5.2 yet is because of this: https://github.com/unoplatform/uno.extensions/issues/2412)
```
Failed to restore D:\Projects\DMXCore\100\Software\Shared\Shared.csproj (in 4 ms).
NuGet package restore failed. Please see Error List window for detailed warnings and errors.
1>D:\Projects\DMXCore\100\Software\Shared\Shared.csproj : error NU1012: Platform version is not present for one or more target frameworks, even though they have specified a platform: net8.0-desktop
1>Done building project "Shared.csproj" -- FAILED.
1>C:\Program Files\dotnet\sdk\8.0.303\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.TargetFrameworkInference.targets(117,5): error NETSDK1139: The target platform identifier desktop was not recognized.
```
If I change the Uno.Sdk version to 5.2.175 then it builds and runs fine (except for message dialogs).
```
  "msbuild-sdks": {
    "Uno.Sdk": "5.1.104"
  },
```

**Reply by jeromelaban** [reply]
`Uno.Sdk` was introduced in 5.1 and later, though in your case, it is likely that you have not specified `Sdk="Uno.Sdk"` in your `.csproj`.

**Reply by jeromelaban** [reply]
> it works if I just change the global.json to 5.2

Ah, yes, my apologies. While uno.sdk is supported in 5.1, single project is not. https://github.com/unoplatform/uno.extensions/issues/2412 has been fixed, I wonder, which part are you missing?

**Reply by jeromelaban** [reply]
> I think the issue with dialogs is that it’s not back ported to 5.2 yet, or at least I don’t know which pre-version it would be in.

Thanks. This change won't be backported to 5.2, but it will be part of 5.3 which will include uno.extensions 4.. You'll also be able to use uno extensions 4.2.x manually if you need it.

**Reply by HakanL** [reply]
(Sorry I didn't reply in-line, I did it from my phone via email, I guess it didn't keep the thread).
*Maybe a stupid question, but how can I use the 5.3 dev pre-release branch in global.json, I tried to just enter `5.3.0-dev.1726`, but then it fails to load the project.*
Never mind, I figured it out, there is a Uno.Sdk Nuget package. I was going off the documentation here https://platform.uno/docs/articles/upgrading-nuget-packages.html that states that I should use the version from Uno.WinUI, but that doesn't seem to be correct.

**Reply by jeromelaban** [reply]
uno.sdk builds are tracked on this package starting from 5.3: https://www.nuget.org/packages/Uno.Sdk

**Reply by HakanL** [reply]
I tried 5.3.0-dev.1916 and that builds, but the issue with the dialog boxes is still there. Do I need to do something else to get the pre-release of the extensions? Exception:
`System.InvalidOperationException: Trying to set position of the dialog before it is associated with a visual tree. Make sure to use InitializeWithWindow before calling ShowAsync.`

**Reply by jeromelaban** [reply]
Latest pre-release is [5.4.0-dev.135](https://www.nuget.org/packages/Uno.Sdk/5.4.0-dev.135), try with this one and see if it helps. Once 5.3 is published it will contain the changes you need.

**Reply by HakanL** [reply]
Hmm, I must be doing something wrong. I've changed global.json to 5.4.0-dev.135 and it builds. But the dialog is still not working, however when I check what packages are dependent then I'm still getting extensions version 4.0.0, see image:
![image](https://github.com/user-attachments/assets/b662c04d-e04c-4c30-9929-aa487146c54c)

**Reply by jeromelaban** [reply, marked as answer]
This like means that you have an explicit version defined in your `Directory.Build.props`. You can remove all the version lines specified there. An update to the documentation is pending for this.

**Reply by HakanL** [reply]
Yes, you're absolutely correct, in fact I even had some in my Directory.Packages.props that I put in there as an attempt to get single-project to work with 5.1, forgot to remove them. Now I'm getting the latest and dialogs are working again. Thanks for all your help, sorry about all the questions, it's tricky to know how to interpret the errors and understand where it's coming from in MSBuild land.

**Reply by jeromelaban** [reply]
No worries and thanks for the troubleshooting, that's why we're simplifying our solutions to avoid all theses errors.

**Comment by HakanL** [comment]
Hmm, no, I definitely have the sdk reference in the csproj, it works if I just change the global.json to 5.2, no other change. But I can test some more with it in a sample project, just thought if you had any suggestions on what may trigger the issue. It seems to be during the restore stage (dotnet restore gives me the same error as VS).

**Comment by HakanL** [comment]
Aha, I didn’t realize they were two separate features in different releases, then it makes sense.
I think the issue with dialogs is that it’s not back ported to 5.2 yet, or at least I don’t know which pre-version it would be in.
